###IMPORTING DEPENDENCIES

In [22]:
import numpy as np #for arrays
import pandas as pd #for reading csv files to work upon
from sklearn.model_selection import train_test_split #for training and testing the data
from sklearn.linear_model import LogisticRegression #chossing this model for data
from sklearn.metrics import accuracy_score #for checomg the accuracy of model

In [23]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [24]:
# loading dataset to pandas dataframe
credit_card_data = pd.read_csv("/content/drive/MyDrive/creditcard.csv")

In [25]:
#first five rows of dataset
credit_card_data.head(5)

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [26]:
#last five rows of dataset 
credit_card_data.tail(5)

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

In [27]:
credit_card_data.shape

(284807, 31)

In [28]:
#checking number of missing values in column
credit_card_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [29]:
credit_card_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [30]:
credit_card_data.describe()

Time            V1            V2            V3            V4  \
count  284807.000000  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean    94813.859575  1.168375e-15  3.416908e-16 -1.379537e-15  2.074095e-15   
std     47488.145955  1.958696e+00  1.651309e+00  1.516255e+00  1.415869e+00   
min         0.000000 -5.640751e+01 -7.271573e+01 -4.832559e+01 -5.683171e+00   
25%     54201.500000 -9.203734e-01 -5.985499e-01 -8.903648e-01 -8.486401e-01   
50%     84692.000000  1.810880e-02  6.548556e-02  1.798463e-01 -1.984653e-02   
75%    139320.500000  1.315642e+00  8.037239e-01  1.027196e+00  7.433413e-01   
max    172792.000000  2.454930e+00  2.205773e+01  9.382558e+00  1.687534e+01   

                 V5            V6            V7            V8            V9  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   9.604066e-16  1.487313e-15 -5.556467e-16  1.213481e-16 -2.406331e-15   
std    1.380247e+00  1.332271e+00  1.237094e+00  1.194353e+00  1.098632e+00   
min   -1.137433e+02 -2.616051e+01 -4.355724e+01 -7.321672e+01 -1.343407e+01   
25%   -6.915971e-01 -7.682956e-01 -5.540759e-01 -2.086297e-01 -6.430976e-01   
50%   -5.433583e-02 -2.741871e-01  4.010308e-02  2.235804e-02 -5.142873e-02   
75%    6.119264e-01  3.985649e-01  5.704361e-01  3.273459e-01  5.971390e-01   
max    3.480167e+01  7.330163e+01  1.205895e+02  2.000721e+01  1.559499e+01   

       ...           V21           V22           V23           V24  \
count  ...  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   ...  1.654067e-16 -3.568593e-16  2.578648e-16  4.473266e-15   
std    ...  7.345240e-01  7.257016e-01  6.244603e-01  6.056471e-01   
min    ... -3.483038e+01 -1.093314e+01 -4.480774e+01 -2.836627e+00   
25%    ... -2.283949e-01 -5.423504e-01 -1.618463e-01 -3.545861e-01   
50%    ... -2.945017e-02  6.781943e-03 -1.119293e-02  4.097606e-02   
75%    ...  1.863772e-01  5.285536e-01  1.476421e-01  4.395266e-01   
max    ...  2.720284e+01  1.050309e+01  2.252841e+01  4.584549e+00   

                V25           V26           V27           V28         Amount  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  284807.000000   
mean   5.340915e-16  1.683437e-15 -3.660091e-16 -1.227390e-16      88.349619   
std    5.212781e-01  4.822270e-01  4.036325e-01  3.300833e-01     250.120109   
min   -1.029540e+01 -2.604551e+00 -2.256568e+01 -1.543008e+01       0.000000   
25%   -3.171451e-01 -3.269839e-01 -7.083953e-02 -5.295979e-02       5.600000   
50%    1.659350e-02 -5.213911e-02  1.342146e-03  1.124383e-02      22.000000   
75%    3.507156e-01  2.409522e-01  9.104512e-02  7.827995e-02      77.165000   
max    7.519589e+00  3.517346e+00  3.161220e+01  3.384781e+01   25691.160000   

               Class  
count  284807.000000  
mean        0.001727  
std         0.041527  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000  

[8 rows x 31 columns]

In [31]:
#distribution of legit & fraudulent transaction
credit_card_data['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

This dataset is highly unbalanced

As we can see there is more than 99% of data that is legit transaction so we cannot make clear to our machine learning model & it couldn't recognize which are fraudaulent data because we have very less data points on that particular case.
so everytime we will input data it will show as legit transaction. So therefore it is unbalanced dataset.


0 --> Legit Transaction

1 --> Fraudulent Transaction

In [32]:
 # separating data for analysis
legit = credit_card_data[credit_card_data.Class == 0]
fraud = credit_card_data[credit_card_data.Class == 1]


In [33]:
print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


In [34]:
#statistical measures of data
legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [35]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [36]:
#compare the values for both transaction
credit_card_data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

Under-Sampling

Build sample dataset containing similar distribution of normal transaction and fraudulent transaction

number of fraudulent transaction --> 492


In [37]:
legit_sample = legit.sample(n=492) #provides random sample 

concatinating two DataFrame

In [38]:
new_dataset = pd.concat([legit_sample,fraud],axis=0)

#axis = 0 specifies it should be added row wise

In [39]:
new_dataset.head(5)

Time        V1        V2        V3        V4        V5        V6  \
161644  114386.0 -1.667452  0.838901 -0.752086 -3.894523  3.068133  3.028741   
268258  163134.0 -0.230261  0.958777  0.077597 -0.835338  0.664087 -0.560585   
235992  148628.0 -0.563674 -0.253162  0.605047 -1.010878 -0.678905  0.481694   
25315    33554.0  1.027341 -0.278164  0.587195  0.661251 -0.340992  0.606581   
270175  163950.0 -0.640625  1.409063 -1.362859 -1.246987  1.024760 -1.271388   

              V7        V8        V9  ...       V21       V22       V23  \
161644  0.798818  0.908386  1.101122  ... -0.412350 -0.697680 -0.313552   
268258  0.869069 -0.015504 -0.126147  ... -0.252997 -0.499122  0.027056   
235992 -1.063766 -2.552228 -1.475811  ... -1.647566 -0.722814 -0.460308   
25315  -0.401929  0.382128  0.422348  ... -0.183180 -0.454513  0.108798   
270175  1.244314  0.097518 -0.612902  ...  0.279516  0.918182 -0.271192   

             V24       V25       V26       V27       V28  Amount  Class  
161644  0.643463  1.283951 -1.257221  0.075928 -0.288702    7.30      0  
268258 -0.398759 -0.431303  0.150339  0.355943  0.145782    2.67      0  
235992 -0.547202  1.361440  0.456248 -0.074625  0.168180  169.80      0  
25315  -0.285382  0.110871  0.222245  0.000716  0.003707   39.95      0  
270175 -0.439002 -0.201501  0.104490  0.380239  0.240246    7.70      0  

[5 rows x 31 columns]

In [40]:
new_dataset.tail(5)

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

In [42]:
new_dataset['Class'].value_counts()

0    492
1    492
Name: Class, dtype: int64

In [44]:
new_dataset.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      92952.266260  0.130098  0.004725  0.124741 -0.005318  0.004972   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.128568 -0.050677  0.101695 -0.003472  ... -0.056535 -0.044012   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0      0.000112  0.021431  0.027368  0.003141 -0.011036 -0.000768 -0.013459   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       66.460854  
1      122.211321  

[2 rows x 30 columns]

Splitting data into features & targets

Targets are nothing but 0 & 1

Features are this v1,v2,...,v28,amount

In [45]:
x = new_dataset.drop(columns='Class',axis=1)  #droping column so axis is set to 1
y=new_dataset['Class']

In [46]:
print(x)

            Time        V1        V2        V3        V4        V5        V6  \
161644  114386.0 -1.667452  0.838901 -0.752086 -3.894523  3.068133  3.028741   
268258  163134.0 -0.230261  0.958777  0.077597 -0.835338  0.664087 -0.560585   
235992  148628.0 -0.563674 -0.253162  0.605047 -1.010878 -0.678905  0.481694   
25315    33554.0  1.027341 -0.278164  0.587195  0.661251 -0.340992  0.606581   
270175  163950.0 -0.640625  1.409063 -1.362859 -1.246987  1.024760 -1.271388   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  .

In [47]:
print(y)

161644    0
268258    0
235992    0
25315     0
270175    0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


Now we will split this features and targets into training data and test data

For this we will use train test split which is imported from sklearn.model_selection

we need to mention four variables...

In [48]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2,stratify=y,random_state=2)

#test size represents the amount of testing data we want
#so 80% of data goes to training data and 20% goes to testing data
#stratify evenly distributes y into training and testing data


In [49]:
print(x.shape,x_train.shape,x_test.shape)

(984, 30) (787, 30) (197, 30)


so here we can see that in x there were 984 values and out of that 787 values gone to x_train and 197 values goes to x_test

**Model Training**

Logistic Regression

so generally we use logistic regression for binary classification problems

In [50]:
model = LogisticRegression()

In [51]:
# Training logistic regression model with training data
model.fit(x_train,y_train) 
#so as we know x_train contains all the features of training data and y_train contains corresponding labels
#labels are nothing but o & 1


LogisticRegression()

Model Evaluation

Accuracy Score

In [52]:
# first we are finding accuracy score on our training data
x_train_prediction = model.predict(x_train)
training_data_accuracy = accuracy_score(x_train_prediction,y_train)


In [53]:
print('Accuracy on Training Data : ',training_data_accuracy)

Accuracy on Training Data :  0.9453621346886912


In [55]:
# we got accuracy score of 94.1% which is very good accuracy .
# so 94.1% means out of 100 predictions our model can predict almost 94 predictions correct

In [56]:
# testing data is the values which our model has not seen earlier so based on that also evaluation is important

In [58]:
# Accuracy on Test Data
x_test_prediction = model.predict(x_test)
test_data_accuracy = accuracy_score(x_test_prediction,y_test)

In [59]:
print('Accuracy on Testing Data : ',test_data_accuracy)

Accuracy on Testing Data :  0.9289340101522843


In [62]:
# here we are getting almost same accuracy as of training data 
# if it would be different then our model would have been overfitted or underfitted 
# let say if accuracy of training data is 94% and accuracy of test data is 50% then it means our model has overfitted with the training data
# let say if accuracy of training data is 50% and accuracy of test data is 94% then it means model is underfit with the training data